File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/ObjTileMapsForge.java
package ch.bailu.aat_lib.service.cache;

import org.mapsforge.core.graphics.TileBitmap;
import org.mapsforge.core.model.Tile;

import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.map.tile.MapTileInterface;
import ch.bailu.aat_lib.preferences.map.SolidTileSize;

public final class ObjTileMapsForge extends ObjTile {
    private static long DEFAULT_SIZE = SolidTileSize.DEFAULT_TILESIZE_BYTES * 4;

    private final AppContext appContext;
    private final Tile tile;

    private final MapTileInterface bitmap;

    private final String themeID;

    public ObjTileMapsForge(String id, AppContext appContext, Tile t, String tID) {
        super(id);
        this.appContext = appContext;
        tile = t;
        themeID = tID;

        bitmap = appContext.createMapTile();

        this.appContext.getServices().getRenderService().lockToRenderer(this);
    }


    public String getThemeID() {
        return themeID;
    }


    @Override
    public void reDownload(AppContext sc) {}

    @Override
    public boolean isLoaded() {
        return bitmap.isLoaded();
    }

    @Override
    public void onDownloaded(String id, String url, AppContext sc) {}

    @Override
    public void onChanged(String id, AppContext sc) {}


    public void onRendered(TileBitmap fromRenderer) {
            bitmap.set(fromRenderer);
            appContext.getBroadcaster().broadcast(
                    AppBroadcaster.FILE_CHANGED_INCACHE,
                    getID());
    }


    @Override
    public void onRemove(AppContext sc) {
        appContext.getServices().getRenderService().freeFromRenderer(ObjTileMapsForge.this);
        bitmap.free();
        super.onRemove(sc);
    }


    @Override
    public long getSize() {
        DEFAULT_SIZE = ObjTile.getSize(bitmap, DEFAULT_SIZE);

        if (isLoaded()) {
            return DEFAULT_SIZE;
        } else {
            return DEFAULT_SIZE * 4;
        }
    }

    @Override
    public TileBitmap getTileBitmap() {
        return bitmap.getTileBitmap();
    }

    public Tile getTile() {
        return tile;
    }



    public static class Factory extends Obj.Factory {
        private final Tile mapTile;
        private final String themeID;

        public Factory(Tile t, String tID) {

            themeID = tID;
            mapTile=t;
        }

        @Override
        public Obj factory(String id, AppContext appContext) {
            return  new ObjTileMapsForge(id, appContext, mapTile, themeID);
        }

    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/ObjTileMapsForge.kt
package ch.bailu.aat_lib.service.cache

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.preferences.map.SolidTileSize
import org.mapsforge.core.graphics.TileBitmap
import org.mapsforge.core.model.Tile

class ObjTileMapsForge(
    id: String,
    private val appContext: AppContext,
    private val tile: Tile,
    @JvmField val themeID: String
) : ObjTile(id) {
    private val bitmap = appContext.createMapTile()

    init {
        appContext.services.getRenderService().lockToRenderer(this)
    }

    override fun reDownload(sc: AppContext) {}

    override fun isLoaded(): Boolean {
        return bitmap.isLoaded()
    }

    override fun onDownloaded(id: String, url: String, sc: AppContext) {}

    override fun onChanged(id: String, sc: AppContext) {}


    fun onRendered(fromRenderer: TileBitmap?) {
        bitmap.set(fromRenderer)
        appContext.broadcaster.broadcast(
            AppBroadcaster.FILE_CHANGED_INCACHE,
            id
        )
    }


    override fun onRemove(sc: AppContext) {
        appContext.services.getRenderService().freeFromRenderer(this@ObjTileMapsForge)
        bitmap.free()
        super.onRemove(sc)
    }


    override fun getSize(): Long {
        DEFAULT_SIZE = getSize(bitmap, DEFAULT_SIZE)

        return if (isLoaded) {
            DEFAULT_SIZE
        } else {
            DEFAULT_SIZE * 4
        }
    }

    override fun getTileBitmap(): TileBitmap? {
        return bitmap.getTileBitmap()
    }

    override fun getTile(): Tile {
        return tile
    }


    class Factory(private val mapTile: Tile, private val themeID: String) : Obj.Factory() {
        override fun factory(id: String, appContext: AppContext): Obj {
            return ObjTileMapsForge(id, appContext, mapTile, themeID)
        }
    }

    companion object {
        private var DEFAULT_SIZE = (SolidTileSize.DEFAULT_TILESIZE_BYTES * 4).toLong()
    }
}
